### ISSUE TITLE
Feature request: Anonymize IP-Addresses (legal issues in EU region)

### ISSUE THREAD

**Issue by david-uc** [mautic/mautic#3445]
| Q  | A
| ---| ---
| Bug report? | 
| Feature request? | Y
| Enhancement? | 

## Description:
It would be great to have a switch somewhere in the dashboard to anonymize all IP addresses by the last octet, just like Google Analytics does with their AnonIP function.

This would decrease the accuracy of the users location by only a small percentage and help all EU users to actually use Mautic legally. Within the EU we cannot save or display full IP addresses due to privacy laws.

## Caveats:

1. The visitor's IP (192.168.0.144) would be either displayed (and saved) as 192.168.0.### or as a hash value (ie. md5 -> 2d992110258ce8929595256a90dc9b65)
2. The IP lookup can be made with either the anonymized IP (192.168.0.###) or could be made with the real IP and later on saved as an anonymized IP (ie. md5-hash).
3. Every visitors IP address needs to get hashed and checked against the DB (just like with IP addresses now).

Currently we (in Europe) need to hack the core of every version to use Mautic legally.

**Comment by merlin0** [comment]
+1

**Comment by RCheesley** [comment]
Sounds like a sensible idea, definitely think that there should be a way to change between showing full address and showing partial as suggested, in case it's turned on in error.

**Comment by pavel-janicek** [comment]
Can you please elaborate more? In Czech Republic (EU member state) the law is, that you have to register self as someone who stores user data. And so far thats all I know about legal issues concerning collection of user data, including full IP address

**Comment by uberconversion** [comment]
I am not a lawyer and can't really go into details. But since many of us do host our servers in non-EU datacenters saving the real IP addresses together with personally identifiable information can be tricky, if not even illegal (due to privacy protection laws).

It may or may not be OK in some EU countries, but in some highly competitive online markets I know of competitors who will not blink an eye to sue you for using something like mautic. You better lawyer up and be damn sure that what you are doing is 100% legal in your country, with the servers location, the website you integrate the tracking snippet and the target market your site is operating in and for.

I'd personally sleep way better knowing no real IP addresses are saved in my DB and thus I can circumvent the whole legalese.

**Comment by thomeudt** [comment]
+1

We are also hacking the core bundle right now to be compliant with German law.

Our workaround is to mask the ip_address in the respective setter to ensure the IP is anonymized before persisting:

```
    public function setIpAddress($ipAddress)
    {
        $this->ipAddress = preg_replace('/(?!\d{1,3}\.\d{1,3}\.)\d/', '*', $ipAddress);

        return $this;
    }

```

It's not a pretty way of doing this. Should be possible to toggle this behavior using the GUI interface.

**Comment by Woeler** [comment]
Merged and planned for 2.15.0 https://github.com/mautic/mautic/pull/6121
